project/scheduling/Schedule.agda

```diff
@@ -0,0 +1,90 @@
+module Schedule where
+open import Relation.Binary.PropositionalEquality
+open ≡-Reasoning
+open import Data.Bool using (Bool ; true ; false ; if_then_else_ ; _∨_ ; _∧_)
+open import Data.List using (List ; [] ; _∷_ ; map ; all ; any ; and ; or )
+open import Data.List.Relation.Unary.AllPairs using (AllPairs; allPairs?)
+open import Data.List.Relation.Unary.All using (All)
+open import Data.Maybe using (Maybe ; nothing ; just)
+open import Data.Sum using (_⊎_)
+open import Level using (0ℓ)
+open import Relation.Unary using (Pred)
+open import Relation.Binary
+open import Relation.Nullary
+open import Relation.Nullary.Decidable using (True; False; ⌊_⌋; toWitness; fromWitness)
+open import Relation.Nullary.Negation
+open import Relation.Nullary.Product
+open import Relation.Nullary.Sum
+open import Data.Empty
+open import Data.Product using (_×_)
+open import Interval
+open import Pair
+open import AcyclicGraph
+open import Agda.Builtin.Unit
+open import DecidableEquality
+
+open DecEq {{...}}
+
+Task = AcyclicGraphNode ⊤ Interval
+TaskGraph = AcyclicGraph ⊤ Interval
+
+data _↝⇒<<_ : Rel Task 0ℓ where
+  dependency-implies-ordering : {l r : Task} → {proof : True (((l ↝? r) ×-dec (get-val l <<? get-val r)) ⊎-dec ¬? (l ↝? r))} → l ↝⇒<< r
+
+↝⇒<<⇒p : {l r : Task} → l ↝⇒<< r → ((l ↝ r) × (get-val l << get-val r) ⊎ (¬ l ↝ r))
+↝⇒<<⇒p (dependency-implies-ordering {proof = p}) = toWitness p
+
+_↝⇒<<?_ : Decidable {A = Task} _↝⇒<<_
+l ↝⇒<<? r with  ((l ↝? r) ×-dec (get-val l <<? get-val r)) ⊎-dec ¬? (l ↝? r)
+...          | yes p = yes (dependency-implies-ordering {proof = fromWitness p})
+...          | no ¬p = no lem
+  where
+  lem : (x : l ↝⇒<< r) → ⊥
+  lem (dependency-implies-ordering {proof = p}) = contradiction (toWitness p) ¬p
+
+
+data _↝⇔<<_ : Rel Task 0ℓ where
+  bidierctional-dependency-implies-ordering : {l r : Task} → {proof : True ((l ↝⇒<<? r) ×-dec (r ↝⇒<<? l))} → l ↝⇔<< r
+
+_↝⇔<<?_ : Decidable {A = Task} _↝⇔<<_
+l ↝⇔<<? r with ((l ↝⇒<<? r) ×-dec (r ↝⇒<<? l))
+...          | yes p = yes (bidierctional-dependency-implies-ordering {proof = fromWitness p})
+...          | no ¬p = no lem
+  where
+  lem : (x : l ↝⇔<< r) → ⊥
+  lem (bidierctional-dependency-implies-ordering {proof = p}) = contradiction (toWitness p) ¬p
+
+record Schedule : Set where
+  constructor s[_]s
+  field
+    schedule : AcyclicGraph ⊤ Interval
+    {no-interval-intersection} : True (NoneIntersect? (map get-val (all-nodes schedule)))
+    {forall-dependencies-imply-ordering} : True (allPairs? _↝⇔<<?_ (all-nodes schedule))
+
+schedule⇒no-interval-intersection : (s : Schedule) → NoneIntersect (map get-val (all-nodes (Schedule.schedule s)))
+schedule⇒no-interval-intersection s = toWitness (Schedule.no-interval-intersection s)
+
+schedule⇒forall-dependencies-imply-ordering : (s : Schedule) → AllPairs _↝⇔<<_ (all-nodes (Schedule.schedule s))
```

Comment:
This, along with the function above, are the two constraints that the `Schedule` object enforces at compile-time.